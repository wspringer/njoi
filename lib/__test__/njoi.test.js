const Joi = require("@hapi/joi");

const schema1 = Joi.object()
  .keys({
    a: Joi.string()
      .required()
      .description(
        "Some string alkdsf jlkasj dfjklasj dfklj aklsj fkljaklsdfj klajs fklj aklsj fklja sklfjkl asjdfklj asklj klajsklfj aklsj dfklaskldfj klaj skldf jf"
      )
      .example("yay"),
    b: Joi.number().description("Some number").default(3),
    c: Joi.boolean().description("Some boolean"),
    d: Joi.array().items(Joi.string()),
    e: Joi.array().items(
      Joi.object()
        .keys({
          first: Joi.string(),
        })
        .label("Person")
    ),
    f: Joi.string().valid("a", "b", "c"),
    type: Joi.string().valid("condo", "appartment", "mansion"),
  })
  .description(
    "Impossible long description with way more text than should be allowed on a single line, but adding it anyhow to make sure wrapping is working. Impossible long description with way more text than should be allowed on a single line, but adding it anyhow to make sure wrapping is working. "
  )
  .label("Simple")
  .unknown(true);

describe("njoi", () => {
  const njoi = require("../njoi");
  it("should a sensible description", () => {
    expect(njoi.jsonish({ comments: true })(schema1)).toMatchSnapshot();

    expect(njoi.markdown()(schema1)).toMatchSnapshot();
  });

  it("should allow you to generate jsdoc", () => {
    const schema = schema1.keys({
      g: Joi.object()
        .pattern(
          Joi.string()
            .valid("a", "b", "c")
            .description("The first three characters.")
            .label("Character"),
          Joi.object()
            .keys({
              top: Joi.number().description("The top coordinate"),
              left: Joi.number().description("The left coordinate"),
            })
            .label("Location")
        )
        .description("The coordinates indexed by characters")
        .label("CoordinateHolder"),
      h: Joi.alternatives().try(
        Joi.string(),
        Joi.object({
          first: Joi.string(),
          second: Joi.number(),
        })
      ),
    });

    console.info(JSON.stringify(schema.describe(), null, 2));

    console.info(njoi.jsonish({ comments: false })(schema));
    expect(njoi.jsdoc()(schema)).toMatchSnapshot();
    console.info(njoi.jsdoc()(schema));
    console.info(njoi.jsonish({ comments: false })(schema));
    console.info(njoi.markdown()(schema));
  });

  it("should treat strings with defaults as an other type", () => {
    const schema = Joi.object({ foo: Joi.string().default("bar") });
    console.info(njoi.jsonish()(schema));
  });
});
