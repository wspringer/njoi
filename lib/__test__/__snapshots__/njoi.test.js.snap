// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`njoi should a sensible description 1`] = `
"{
  /**
   * Some string alkdsf jlkasj dfjklasj dfklj aklsj 
   * fkljaklsdfj klajs fklj aklsj fklja sklfjkl 
   * asjdfklj asklj klajsklfj aklsj dfklaskldfj klaj 
   * skldf jf
   */
  a: \\"yay\\",

  /**
   * Some number
   */
  b: <number>,

  /**
   * Some boolean
   */
  c: <boolean>,

  d: [<string>, ...],

  e: [{
    first: <string>,
  }, ...],

  f: <string>,

  type: <string>,
}"
`;

exports[`njoi should allow you to generate jsdoc 1`] = `
"{
  a: \\"yay\\",

  b: <number>,

  c: <boolean>,

  d: [<string>, ...],

  e: [{
    first: <string>,
  }, ...],

  f: <string>,

  type: <string>,

  g: {
    a|b|c: {
      top: <number>,

      left: <number>,
    },
  },
}"
`;

exports[`njoi should allow you to generate jsdoc 2`] = `
"/**
 * @typedef {Object} Simple
 * @property {string} a - Some string alkdsf jlkasj dfjklasj dfklj aklsj fkljaklsdfj klajs fklj aklsj fklja sklfjkl asjdfklj asklj klajsklfj aklsj dfklaskldfj klaj skldf jf
 * @property {number} [b=3] - Some number
 * @property {boolean} [c] - Some boolean
 * @property {string[]} [d]
 * @property {Person[]} [e]
 * @property {('a'|'b'|'c')} [f]
 * @property {('condo'|'appartment'|'mansion')} [type]
 * @property {CoordinateHolder} [g] - The coordinates indexed by characters
 */
/**
 * @typedef {Object} Person
 * @property {string} [first]
 */
/**
 * The coordinates indexed by characters
 * @typedef {Object.<string,Location>} CoordinateHolder - Valid keys: 'a', 'b', 'c'
 */
/**
 * @typedef {Object} Location
 * @property {number} [top] - The top coordinate
 * @property {number} [left] - The left coordinate
 */
"
`;
